content/blog/2018/incremental-traffic-management/index.md

```diff
@@ -0,0 +1,225 @@
+
+{{< text bash >}}
+$ kubectl apply -f samples/bookinfo/platform/kube/bookinfo.yaml
+{{< /text >}}
+
+Now, **configure a new Gateway** that allows access to the reviews service from outside the Istio mesh, a new VirtualService that splits traffic evenly between v1 and v2 of the reviews service, and a set of new DestinationRules that match destination subsets to service versions:
+
+{{< text bash >}}
+$ cat <<EOF | kubectl apply -f -
+apiVersion: networking.istio.io/v1alpha3
+kind: Gateway
+metadata:
+  name: reviews-gateway
+spec:
+  selector:
+    istio: ingressgateway # use istio default controller
+  servers:
+  - port:
+      number: 80
+      name: http
+      protocol: HTTP
+    hosts:
+    - "*"
+---
+apiVersion: networking.istio.io/v1alpha3
+kind: VirtualService
+metadata:
+  name: reviews
+spec:
+  hosts:
+  - "*"
+  gateways:
+  - reviews-gateway
+  http:
+  - match:
+    - uri:
+        prefix: /reviews
+    route:
+    - destination:
+        host: reviews
+        subset: v1
+      weight: 50
+    - destination:
+        host: reviews
+        subset: v2
+      weight: 50
+---
+apiVersion: networking.istio.io/v1alpha3
+kind: DestinationRule
+metadata:
+  name: reviews
+spec:
+  host: reviews
+  subsets:
+  - name: v1
+    labels:
+      version: v1
+  - name: v2
+    labels:
+      version: v2
+  - name: v3
+    labels:
+      version: v3
+EOF
+{{< /text >}}
+
+Finally, **deploy a Pod that you can use for testing** with `curl`:
+
+{{< text bash >}}
+$ kubectl apply -f samples/sleep/sleep.yaml
+{{< /text >}}
+
+## Testing your deployment
+
+Now, you can test different behaviors using the `curl` commands via the sleep Pod. 
+
+The first example is to issue requests to the reviews service using standard Kubernetes service DNS behavior (**note**: [`jq`](https://stedolan.github.io/jq/) is used in the examples below to filter the output from `curl`):
+
+{{< text bash >}}
+$ export SLEEP_POD=$(kubectl get pod -l app=sleep \
+  -o jsonpath={.items..metadata.name})
+$ for i in `seq 3`; do \
+  kubectl exec -it $SLEEP_POD curl http://reviews:9080/reviews/0 | \
+  jq '.reviews|.[]|.rating?'; \
+  done
+{{< /text >}}
+
+{{< text json >}}
+{
+  "stars": 5,
+  "color": "black"
+}
+{
+  "stars": 4,
+  "color": "black"
+}
+null
+null
+{
+  "stars": 5,
+  "color": "red"
+}
+{
+  "stars": 4,
+  "color": "red"
+}
+{{< /text >}}
+
+Notice how we’re getting responses from all three versions of the reviews service (`null` is from reviews v1 which doesn’t have ratings) and not getting the even split across v1 and v2. This is expected behavior because the `curl` command is using Kubernetes service load balancing across all three versions of the reviews service. In order to access the reviews 50/50 split we need to access the service via the ingress Gateway:
+
+{{< text bash >}}
+$ for i in `seq 4`; do \ 
+  kubectl exec -it $SLEEP_POD curl http://istio-ingressgateway.istio-system/reviews/0 | \
```

Comment:
This seems a little hacky. If you were going to do something like this in production, you would probably want to configure DNS for a real host instead of configuring host `*` and then just calling the ingressgateway service directly. I wonder if you could use the new coredns addon that's used for multi-cluster config: https://preliminary.istio.io/docs/setup/kubernetes/multicluster-install/gateways/#configure-dns

Of course that starts to move away from the minimal Istio install theme of this article.

Reply:
Agreed, this moves in the direction of increased complexity so keeping it simpler for this post.